Mypy annotations

stinovlas opened the pull request:
Closes #4.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/stinovlas/django-pain/5)
<!-- Reviewable:end -->

stinovlas committed:
Add mypy check

stinovlas committed:
Add type annotations

ziima commented on tox.ini:
```diff
@@ -36,6 +36,7 @@ commands =
     isort --recursive --check-only --diff django_pain
     flake8 --format=pylint --show-source django_pain
     pydocstyle django_pain
+    mypy -p django_pain
```
`--package`

stinovlas commented:
*[tox.ini, line 39 at r1](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCTpK6TTiie9usYCWUd-r1-39:-LCYq1kL7BmwMgpjwEqb:b-t790f0) ([raw file](https://github.com/stinovlas/django-pain/blob/c973c0db2d7291e9a13aa090965710d5f2a82f4b/tox.ini#L39)):*
<details><summary><i>Previously, ziima (Vlastimil Zíma) wrote…</i></summary><blockquote>

`--package`
</blockquote></details>

Fixed :shipit:.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

ziima commented:
Reviewed 1 of 8 files at r1, 2 of 6 files at r2, 3 of 5 files at r3.
Review status: 6 of 8 files reviewed at latest revision, all discussions resolved.

---

*[django_pain/management/commands/import_payments.py, line 5 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtYrGCcfh26HvUg1W:-LCYtYrGCcfh26HvUg1X:bom5jl7) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L5)):*
> ```Python
> from typing import Sequence
> 
> import django.utils.module_loading as module_loading
> ```

I don't get the change of import.

---

*[django_pain/management/commands/import_payments.py, line 30 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtjSz8f0baCMUdrRL:-LCYtjSz8f0baCMUdrRM:buqat79) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L30)):*
> ```Python
>         parser = module_loading.import_string(options['parser'])()  # type: AbstractBankStatementParser
>         if not isinstance(parser, AbstractBankStatementParser):
>             raise CommandError('Parser argument has to be subclass of AbstractBankStatementParser.')
> ```

I don't think we need strong type check.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

stinovlas commented:
Review status: 6 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[django_pain/management/commands/import_payments.py, line 5 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtYrGCcfh26HvUg1W:-LCYv9pi4eQdf2mq3upB:b-1u24ih) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L5)):*
<details><summary><i>Previously, ziima (Vlastimil Zíma) wrote…</i></summary><blockquote>

I don't get the change of import.
</blockquote></details>

Well, `mypy` was complaining before I did this. But it's not complaining now, so I reverted this change.

---

*[django_pain/management/commands/import_payments.py, line 30 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtjSz8f0baCMUdrRL:-LCYvIpID0sFMJ7VrdJk:b-lttzwc) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L30)):*
<details><summary><i>Previously, ziima (Vlastimil Zíma) wrote…</i></summary><blockquote>

I don't think we need strong type check.
</blockquote></details>

What do you mean by that? Python itself is strongly typed. If you mean that we don't need explicit check for argument's class (and customized error), I disagree.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

ziima commented:
Review status: 6 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[django_pain/management/commands/import_payments.py, line 30 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtjSz8f0baCMUdrRL:-LCZ0FvJ8u3jF5UFOvCV:baampre) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L30)):*
<details><summary><i>Previously, stinovlas (Jan Musílek) wrote…</i></summary><blockquote>

What do you mean by that? Python itself is strongly typed. If you mean that we don't need explicit check for argument's class (and customized error), I disagree.
</blockquote></details>

Python's interface is based on duck-typing, so any explicit type checks are against this idea.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

stinovlas commented:
*[django_pain/management/commands/import_payments.py, line 30 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtjSz8f0baCMUdrRL:-LCZ14N02MextFKDupeT:b-wkewrh) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L30)):*
<details><summary><i>Previously, ziima (Vlastimil Zíma) wrote…</i></summary><blockquote>

Python's interface is based on duck-typing, so any explicit type checks are against this idea.
</blockquote></details>

Well, I didn't say, that it's not dynamic. Dynamic language and strong typing can exist together perfectly well. We agreed on doing static type checks. Normally, the types are inferred from the code itself, which can't be done in this case, because return type of `import_string` is just unguessable. I'm just hinting the `mypy` to improve the static type checking. As for checking the user input, I believe that it's a good call in this particular case.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

stinovlas commented:
*[django_pain/management/commands/import_payments.py, line 30 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtjSz8f0baCMUdrRL:-LCbjFv1EOD4KRzT7oPJ:b47uxxh) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L30)):*
<details><summary><i>Previously, stinovlas (Jan Musílek) wrote…</i></summary><blockquote>

Well, I didn't say, that it's not dynamic. Dynamic language and strong typing can exist together perfectly well. We agreed on doing static type checks. Normally, the types are inferred from the code itself, which can't be done in this case, because return type of `import_string` is just unguessable. I'm just hinting the `mypy` to improve the static type checking. As for checking the user input, I believe that it's a good call in this particular case.
</blockquote></details>

@ziima: OK. Let's put it another way. It's my design decision to check the type of the class supplied by user. It's not clashing with functionality and it is tested on its own. I'm really sure that I want to leave that check where it is. As an author, I'm entitled to do this type of design decisions. 

If you have some highly relevant reason why the check should not be here, please point it out. I'm sorry, but "against the idea of duck-typing" is not enough. Supporting type annotations and at the same time banishing explicit type checks is a little hypocritic.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

stinovlas commented:
*[django_pain/management/commands/import_payments.py, line 30 at r3](https://reviewable.io/reviews/stinovlas/django-pain/5#-LCYtjSz8f0baCMUdrRL:-LChDJpk6w5SIIB5BqfB:b4c3ji5) ([raw file](https://github.com/stinovlas/django-pain/blob/cdb900dfd4ae601bf433b650b1dcf9d4b4451ff7/django_pain/management/commands/import_payments.py#L30)):*
<details><summary><i>Previously, stinovlas (Jan Musílek) wrote…</i></summary><blockquote>

@ziima: OK. Let's put it another way. It's my design decision to check the type of the class supplied by user. It's not clashing with functionality and it is tested on its own. I'm really sure that I want to leave that check where it is. As an author, I'm entitled to do this type of design decisions. 

If you have some highly relevant reason why the check should not be here, please point it out. I'm sorry, but "against the idea of duck-typing" is not enough. Supporting type annotations and at the same time banishing explicit type checks is a little hypocritic.
</blockquote></details>

I updated the check as we discussed and rebased this branch on master.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

ziima commented:
Reviewed 1 of 5 files at r3.
Review status: 7 of 8 files reviewed at latest revision, all discussions resolved.

---

*[django_pain/management/commands/import_payments.py, line 28 at r5](https://reviewable.io/reviews/stinovlas/django-pain/5#-LChFgOM5cQn4LpMuJUv:-LChFgOM5cQn4LpMuJUw:b-wtk1tf) ([raw file](https://github.com/stinovlas/django-pain/blob/464b76e14a3b0ef1189dc495d31e02767a580d4d/django_pain/management/commands/import_payments.py#L28)):*
> ```Python
>         self.options = options
> 
>         parser_module = module_loading.import_string(options['parser'])
> ```

It's a class, not a module.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

stinovlas commented:
*[django_pain/management/commands/import_payments.py, line 28 at r5](https://reviewable.io/reviews/stinovlas/django-pain/5#-LChFgOM5cQn4LpMuJUv:-LChH7dG208C7JSwRrrW:bsj646i) ([raw file](https://github.com/stinovlas/django-pain/blob/464b76e14a3b0ef1189dc495d31e02767a580d4d/django_pain/management/commands/import_payments.py#L28)):*
<details><summary><i>Previously, ziima (Vlastimil Zíma) wrote…</i></summary><blockquote>

It's a class, not a module.
</blockquote></details>

Corrected.

---


*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5)*
<!-- Sent from Reviewable.io -->

ziima commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/stinovlas/django-pain/5#-:-LChKgk25yiFIZp5BDvY:bnfp4nl)*
<!-- Sent from Reviewable.io -->
